### PR TITLE
Be/pytests without solution

### PR DESCRIPTION
**Pull request by beccaelenzil**
I removed postman tests and added pytests. Some of the pytests for 400 level responses specify exactly what the responses need to look like, where others are more open ended and just specify what the response code needs to be.

I didn't add anything to the Wave instructions about the pytests. I realize I still need to add to the __init__ a configuration for the tests and instructions to create the test database.

### CODE REVIEW

File: ada-project-docs/setup.md [hunk 1]
```diff
@@ -54,8 +53,8 @@ Create two environment variables that will hold your database URLs.
 Your `.env` may look like this:
 
 ```
-SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/task_list_api_development
-SQLALCHEMY_TEST_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/task_list_api_test
+SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/video_store_api_development
+SQLALCHEMY_TEST_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/video_store_api_test
 ```
 
 ## Run `$ flask db init`
```
Comment (alope107) [inline ada-project-docs/setup.md:60]: We should add a section here noting that they will need to load the dotenv and set the app config in app.__init__.py. We can either spell it out for them step by step or tell them to refer to Learn to review how to do this.

**Comment by alope107** [comment]
Given that we're adding a many-to-many lesson, do we still want the hints.md file?

**Comment by alope107** [comment]
Should we update the necessary dependency versions to get rid of the dependabot warning?

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        "name": CUSTOMER_NAME,
+        "phone": CUSTOMER_PHONE,
+        "postal_code": CUSTOMER_POSTAL_CODE
+    })
+
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 201
+    assert response_body["name"] == CUSTOMER_NAME
```
Comment (alope107) [inline tests/test_wave_01.py]: The wave_01.md file says that the only value required to be returned in the response is the id. We should either change the documentation to say to return more fields, or change this test to only check for id in the response.

File: tests/test_wave_01.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    assert response_body["id"] == CUSTOMER_ID
+    assert response_body["phone"] == CUSTOMER_PHONE
+    assert response_body["postal_code"] == CUSTOMER_POSTAL_CODE
+
+    new_customer = Customer.query.get(1)
+
+    assert new_customer
+    assert new_customer.name == CUSTOMER_NAME
```
Comment (alope107) [inline tests/test_wave_01.py:271]: Should we check the phone and postal code here as well?

File: tests/test_wave_01.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    print("********", response_body)
+
+    # Assert
+    assert response.status_code == 200
+    assert response_body["name"] == f"Updated ${CUSTOMER_NAME}"
+    assert response_body["phone"] == f"Updated ${CUSTOMER_PHONE}"
+    assert response_body["postal_code"] == f"Updated ${CUSTOMER_POSTAL_CODE}"
+
+    customer = Customer.query.get(1)
+    assert customer.name == f"Updated ${CUSTOMER_NAME}"
```
Comment (alope107) [inline tests/test_wave_01.py:353]: Should we test the other fields that are updated as well?

File: tests/test_wave_01.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+def test_update_customer(client, one_customer):
+    # Act
+    response = client.put("/customers/1", json={
+        "name": f"Updated ${CUSTOMER_NAME}",
+        "phone": f"Updated ${CUSTOMER_PHONE}",
+        "postal_code": f"Updated ${CUSTOMER_POSTAL_CODE}"
+    })
+    response_body = response.get_json()
+    print("********", response_body)
```
Comment (alope107) [inline tests/test_wave_01.py]: Remove debugging print statement

File: tests/test_wave_01.py [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+def test_update_video(client, one_video):
+    # Act
+    response = client.put("/videos/1", json={
+        "title": "Updated Video Title",
+        "total_inventory": 2,
+        "release_date": "01-01-2021"
+    })
+    response_body = response.get_json()
+    print("********", response_body)
```
Comment (alope107) [inline tests/test_wave_01.py]: Remove debug print statement

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+def client(app):
+    return app.test_client()
+
+@pytest.fixture
+def one_video(app):
+    new_video = Video(
+        title=VIDEO_TITLE, 
+        release_date=VIDEO_RELEASE_DATE,
+        total_inventory=VIDEO_INVENTORY,
+        )
```
Comment (alope107) [inline tests/conftest.py:43]: Add an available_inventory to this fixture

Comment (alope107) [reply tests/conftest.py:43]: Alternatively, direct students to auto-populate the field using a technique similar to: https://stackoverflow.com/questions/36579355/sqlalchemy-set-default-value-of-one-column-to-that-of-another-column

File: tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    db.session.add(new_video)
+    db.session.commit()
+
+@pytest.fixture
+def one_customer(app):
+    new_customer = Customer(
+        name=CUSTOMER_NAME,
+        postal_code=CUSTOMER_POSTAL_CODE,
+        phone=CUSTOMER_PHONE
+    )
```
Comment (alope107) [inline tests/conftest.py:53]: Either add a videos_checked_out value to this fixture or make it clear that customers should have a default of 0 videos checked out.

File: tests/test_wave_02.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+    response = client.post("/rentals/check-in", json={
+        "customer_id": 1,
+        "video_id": 1
+    })
+
+    response_body = response.get_json()
+
+    assert response.status_code == 400
+    assert response_body == {"message": "No outstanding rentals for customer # 1 and video 1"}
```
Comment (alope107) [inline tests/test_wave_02.py]: Consider removing the '#', it's inconsistent with the rest of the expected responses.

File: tests/test_wave_02.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+    #Assert
+    assert response.status_code == 200
+
+def test_can_delete_video_with_rental(client, one_checked_out_video):
+    # Act
+    response = client.delete("/videos/1")
+
+    #Assert
+    assert response.status_code == 200
```
Comment (alope107) [inline tests/test_wave_02.py:194]: Should the customer's videos_checked_out count be altered when a video that's rented out is deleted? Should we check for this?

File: tests/test_wave_02.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+    assert response.status_code == 200
+    assert response_body == []
+
+def test_can_delete_customer_with_rental(client, one_checked_out_video):
+    # Act
+    response = client.delete("/customers/1")
+
+    #Assert
+    assert response.status_code == 200
```
Comment (alope107) [inline tests/test_wave_02.py:187]: Should a video's available_inventory be updated when a customer renting that video is deleted? Should we check for this?

**Comment by alope107** [comment]
This looks great! There's a few minor things I made comments on, but it overall looks really good. Thanks for putting this together, this looks like it was a lot of work! My apologies it took me so long to get this review out.

File: tests/conftest.py (hunk 7 above)
Comment (beccaelenzil) [reply tests/conftest.py:43]: I believe our recommendation for this is to dynamically calculate the available inventory (total inventory - number of rentals for that video).

File: tests/conftest.py (hunk 8 above)
Comment (beccaelenzil) [reply tests/conftest.py:53]: Similar to available inventory, videos_checked_out can be calculated as the number of rentals that customer has

File: tests/test_wave_01.py (hunk 6 above)
Comment (beccaelenzil) [inline tests/test_wave_01.py:154]:
```suggestion
    response_body = response.get_json()
```

File: tests/test_wave_01.py [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+    })
+
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 201
+    assert response_body["name"] == CUSTOMER_NAME
+    assert response_body["id"] == CUSTOMER_ID
+    assert response_body["phone"] == CUSTOMER_PHONE
+    assert response_body["postal_code"] == CUSTOMER_POSTAL_CODE
```
Comment (beccaelenzil) [inline tests/test_wave_01.py]:
```suggestion
    assert response.status_code == 201
    assert response_body["id"] == CUSTOMER_ID
```

File: tests/test_wave_01.py (hunk 3 above)
Comment (beccaelenzil) [inline tests/test_wave_01.py:271]:
```suggestion
    assert new_customer
    assert new_customer.name == CUSTOMER_NAME
    assert new_customer.postal_code = CUSTOMER_POSTAL_CODE
    assert new_customer.phone = CUSTOMER_PHONE
```

File: tests/test_wave_01.py (hunk 5 above)
Comment (beccaelenzil) [inline tests/test_wave_01.py:331]:
```suggestion
    response_body = response.get_json()
```

File: tests/test_wave_01.py (hunk 4 above)
Comment (beccaelenzil) [inline tests/test_wave_01.py:353]:
```suggestion
    assert customer.name == f"Updated ${CUSTOMER_NAME}"
    assert customer.phone == f"Updated ${CUSTOMER_PHONE}"
    assert customer.postal_code == f"Updated ${CUSTOMER_POSTAL_CODE}"
    
```

File: tests/test_wave_02.py (hunk 9 above)
Comment (beccaelenzil) [inline tests/test_wave_02.py]:
```suggestion
    assert response_body == {"message": "No outstanding rentals for customer 1 and video 1"}
```

File: tests/test_wave_02.py (hunk 11 above)
Comment (beccaelenzil) [reply tests/test_wave_02.py:187]: Per earlier comment, by not making available inventory a model attribute and dynamically calculating this, we don't need to worry about this

File: tests/test_wave_02.py (hunk 10 above)
Comment (beccaelenzil) [reply tests/test_wave_02.py:194]: Per earlier comment, by not making videos_checked_out_count a model attribute and dynamically calculating this, we don't need to worry about this

**Review by beccaelenzil — COMMENTED** [review]
Thanks for the feedback!

File: README.md [hunk 13]
```diff
@@ -31,57 +27,6 @@ The full functionality is spelled out further in the "Project Directions" files
...
-#### Run the Smoke Tests
-1. Run your server. You'll need a running server open before Postman can reach any of your endpoints.
-1. Explore! There are ways to run the whole collection of tests and ways to run each individual test.
-1. To run a collection of tests:
-    1. Click the blue `Run` button. This will launch the collection runner.
-    1. In the collection runner, scroll down in the center pane and click the blue `Start Test` button
-
 ## How to Complete and Submit
 
 Go through the waves one-by-one and build the features of this API.
```
Comment (beccaelenzil) [inline README.md:34]:
```suggestion
Go through the waves one-by-one and build the features of this API.

Use the Requirements in the Project Directions and the unit tests to guide your development.
```

File: tests/test_wave_01.py [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+    # Act
+    response = client.put("/videos/1", json={
+        "title": "Updated Video Title",
+        "total_inventory": 2,
+        "release_date": "01-01-2021"
+    })
+    response_body = response.get_json()
+    print("********", response_body)
+
+    # Assert
```
Comment (beccaelenzil) [inline tests/test_wave_01.py]:
```suggestion
    response_body = response.get_json()

    # Assert
```

File: tests/test_wave_01.py [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+    # Act
+    response = client.put("/customers/1", json={
+        "name": f"Updated ${CUSTOMER_NAME}",
+        "phone": f"Updated ${CUSTOMER_PHONE}",
+        "postal_code": f"Updated ${CUSTOMER_POSTAL_CODE}"
+    })
+    response_body = response.get_json()
+    print("********", response_body)
+
+    # Assert
```
Comment (beccaelenzil) [inline tests/test_wave_01.py]:
```suggestion
    response_body = response.get_json()

    # Assert
```

**Pull request merged**